c/tskit/tables.c

```diff
@@ -8105,6 +8105,285 @@ tsk_table_collection_subset(
+    }
+
+    // maps relates id in other (index) to new id in self
+    node_map = malloc(other->nodes.num_rows * sizeof(*node_map));
+    individual_map = malloc(other->individuals.num_rows * sizeof(*individual_map));
+    population_map = malloc(other->populations.num_rows * sizeof(*population_map));
+    site_map = malloc(other->sites.num_rows * sizeof(*site_map));
+    if (node_map == NULL || individual_map == NULL || population_map == NULL
+        || site_map == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    memset(node_map, 0xff, other->nodes.num_rows * sizeof(*node_map));
+    memset(individual_map, 0xff, other->individuals.num_rows * sizeof(*individual_map));
+    memset(population_map, 0xff, other->populations.num_rows * sizeof(*population_map));
+    memset(site_map, 0xff, other->sites.num_rows * sizeof(*site_map));
+
+    // nodes, individuals, populations
+    for (k = 0; k < (tsk_id_t) num_nodes; k++) {
+        if (node_mapping[k] != TSK_NULL) {
+            node_map[k] = node_mapping[k];
+        } else {
+            ret = tsk_node_table_get_row(&other->nodes, k, &node);
+            if (ret < 0) {
+                goto out;
+            }
+            new_ind = TSK_NULL;
+            if (node.individual != TSK_NULL) {
+                if (individual_map[node.individual] == TSK_NULL) {
+                    tsk_individual_table_get_row(
+                        &other->individuals, node.individual, &ind);
+                    ret = tsk_individual_table_add_row(&self->individuals, ind.flags,
+                        ind.location, ind.location_length, ind.metadata,
+                        ind.metadata_length);
+                    if (ret < 0) {
+                        goto out;
+                    }
+                    individual_map[node.individual] = ret;
+                }
+                new_ind = individual_map[node.individual];
+            }
+            new_pop = TSK_NULL;
+            if (node.population != TSK_NULL) {
+                // keep same pops if add_populations is False
+                if (!add_populations) {
+                    population_map[node.population] = node.population;
+                }
+                if (population_map[node.population] == TSK_NULL) {
+                    tsk_population_table_get_row(
+                        &other->populations, node.population, &pop);
+                    ret = tsk_population_table_add_row(
+                        &self->populations, pop.metadata, pop.metadata_length);
+                    if (ret < 0) {
+                        goto out;
+                    }
+                    population_map[node.population] = ret;
+                }
+                new_pop = population_map[node.population];
+            }
+            ret = tsk_node_table_add_row(&self->nodes, node.flags, node.time, new_pop,
+                new_ind, node.metadata, node.metadata_length);
+            if (ret < 0) {
+                goto out;
+            }
```

Comment:
Isn't this whole chunk duplicated in subset? Maybe we could add a helper method ``tsk_table_collection_remap_node`` or something which does the drudge work. Not much point in copy-pasting all this code.

Reply:
yeah, that is a good idea. kind of weird helper method, but I guess it doesn't hurt if it's only internal.